CIP (v2) Draft (#514)
input-output-hk/ouroboros-leios: docs/cip/README.md

## Specification

Ouroboros Leios achieves higher transaction throughput by allowing block
producers to create additional blocks alongside the regular Praos chain. These

> supplementary blocks, called **Endorser Blocks (EBs)**, carry extra transactions

Reviewer: ```suggestion
supplementary blocks, called **Endorser Blocks (EBs)**, reference extra transactions
```

Not sure how precise we need to be here.

that would otherwise wait for future slots. To ensure safety, these blocks
undergo committee validation before their transactions become part of the
permanent ledger.

The key insight is that we can maintain Praos' security guarantees while
processing more transactions by carefully managing when and how these additional
blocks are validated and included in the chain.

<div align="center">
  <a name="figure-3" id="figure-3"></a>
  <p name="protocol-overview">
    <img src="images/protocol-overview.svg" alt="Leios Chain Structure">
  </p>

<em>Figure 3: Leios chain structure showing the relationship between Ranking
Blocks, Endorser Blocks, and Certificates</em>

</div>

Figure 3 illustrates the high-level Leios chain structure. When block producers
create Ranking Blocks (RBs), they may also announce Endorser Blocks (EBs)

> containing additional transactions. These EBs undergo committee validation, and

Reviewer: ```suggestion
containing references to additional transactions. These EBs undergo committee validation, and
```
Not sure how precise we need to be here.

   preceding RB).
2. The certificate is valid as defined in
   [Certificate Validation](#certificate-validation).
3. At least $L_\text{vote} + L_\text{diff}$ slots have elapsed since the slot of
   the RB that announced the EB.

This timing constraint ensures certified EBs have sufficient time to diffuse
throughout the network before their transactions are included in the ledger,
preventing chain adoption delays.

<a id="rb-content-rules" href="#rb-content-rules"></a>

> [!Important]
>
> RBs can contain either an EB certificate **or** transactions directly, but
> **not** both. All transactions must be valid against the complete ledger state
> (including certified EBs). If a node lacks any certified EB data, it **must**

> > produce an empty RB to maintain chain validity.

Reviewer: Is this in the latest version of the Google Doc?

#### Timing Constraints

The timing constraints work together to maintain Praos' security assumptions
while enabling higher throughput. These constraints prevent scenarios where
honest nodes would be forced to delay chain adoption due to missing data.

> <a id="network-characteristics"></a>**Network Characteristics**

Author: moved back into pp

The protocol timing is built upon observed network characteristics that describe
how quickly different types of data propagate. All timing parameters target
propagation to ~95% of honest stake, which represents practical network-wide
availability:

<div align="center">
<a name="table-1" id="table-1"></a>

> | Characteristic          |             Symbol              | Description                                                          | Observed Range |

Reviewer: The "observed range" is derived from simulations, not real-life networks.

</div>

<a id="three-phase-protocol"></a>**Timing Phases for Certificate Inclusion**

The certificate inclusion process (Steps 3-5) involves three distinct timing
phases that ensure network-wide EB availability:

<a id="equivocation-detection"></a>

**Phase 1: Equivocation Detection ($3\Delta_\text{hdr}$)**  
This phase occurs at the beginning of $L_\text{vote}$ when an RB announces an
EB. During these initial slots of the voting period, the network detects any
attempts by adversaries to create multiple conflicting blocks for the same slot.
The factor of 3 accounts for the worst-case equivocation scenario:

- $\Delta_\text{hdr}$: First header propagation
- $\Delta_\text{hdr}$: Conflicting header propagation
- $\Delta_\text{hdr}$: Equivocation proof propagation

This prevents adversaries from sending different EBs to different parts of the
network, as honest nodes will refuse to vote if they detect equivocation. Voting
cannot begin until this phase completes.

<a id="voting-period"></a>

**Phase 2: Voting Period ($L_\text{vote}$)**  
Voting begins after equivocation detection completes. The voting period must
accommodate both EB propagation and validation:

> $$L_\text{vote} > \underbrace{\Delta_\text{EB}^{\text{opt}}}_{\text{EB propagation}} + \underbrace{\Delta_\text{cpu}}_{\text{validation time}}$$

Author: removed underbrace

This ensures all honest committee members can participate by having sufficient
time to:

1. Receive the EB from the network
2. Fully validate it (verify signatures, execute scripts, update state)

With a 75% voting threshold, certification guarantees that at least 25% of
honest stake has already received and processed the EB, setting up the network

> assumption for the diffusion period.

Reviewer: ```suggestion
assumption for the diffusion period. This percentage may be adjusted based on empirical observations of realistic networks and community preferences.
```

beyond what is permitted to be included directly in the RB.

<a id="eb-structure" href="#eb-structure"></a>**EB Structure**: EBs have a
simplified structure:

- `transaction_references`: Ordered list of transaction references (transaction
  ids)

> > [!Important] Transaction references **must** maintain their order. When
> > applying an EB to the ledger, transactions are processed sequentially in the
> > order specified.

Reviewer 2: ```suggestion
```

Broken markdown and redundant.

<div align="center">
<a name="table-2" id="table-2"></a>

| Parameter                                                         |       Symbol        |    Units     | Description                                                                                            |                                                 Constraints                                                  | Rationale                                                                                                                                                                                                                                        |
| ----------------------------------------------------------------- | :-----------------: | :----------: | ------------------------------------------------------------------------------------------------------ | :----------------------------------------------------------------------------------------------------------: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> | <a id="delta-hdr" href="#delta-hdr"></a>RB header diffusion bound | $\Delta_\text{hdr}$ |     slot     | Maximum allowed time for RB headers to propagate around the network from the slot they were created in |                           $\Delta_\text{hdr} \geq \Delta_\text{hdr}^{\text{obs}}$                            | Protocol parameter conservatively set based on observed network behavior. Required for equivocation detection timing. The RbHeaderRelay mini-protocol must guarantee delivery within this bound                                                  |

Reviewer 2: Also, I would not mention a concrete mini-protocol name here and instead expect the network section to derive requirements from all the specified behavior and assumptions.

| Maximum Plutus steps per transaction                              |          -          |  step units  | Maximum computational steps allowed for Plutus scripts in a single transaction                         |                                                    $> 0$                                                     | Limits computational resources per transaction to enable higher throughput                                                                                                                                                                       |
| Maximum Plutus memory per transaction                             |          -          | memory units | Maximum memory allowed for Plutus scripts in a single transaction                                      |                                                    $> 0$                                                     | Limits memory resources per transaction to enable higher throughput                                                                                                                                                                              |

<em>Table 2: Leios Protocol Parameters</em>

</div>

> 

Reviewer 2: The fact that we have EB Size constraints below (in this section), but timing constraints in the protocol flow section feels very odd to me.

To me it would make most sense to have timing constraints discussed in the protocol parameter section too

> throughput:
>
> - Longer recovery periods ($L_\text{recover}$) require larger bitmaps to track
>   more potential transactions
> - The worst-case scenario assumes one ranking block per slot during the
>   recovery period (i.e. $f_\text{RB} = 1$)
> - Smaller minimum transaction sizes ($T_\text{min}$) allow more transactions
>   per block, requiring more correction bits
>
> **Implicit Constraint on $L_\text{recover}$**:
>
> The requirement that $S_\text{bitmap} < S_\text{RB}$ creates an upper bound on
> the recovery period:
>
> $$L_\text{recover} < 8 \times T_\text{min}$$
>
> For example, with $T_\text{min} = 55$ bytes[^mainnet-min-tx], this gives
> $L_\text{recover} < 440$ slots. This constraint ensures the bitmap can fit
> within a ranking block even in the worst-case scenario of continuous block
> production.
>
> The critical requirement is that all transaction corrections **must** be
> included before the certificate-active window expires to ensure ledger
> integrity. In extreme cases, the first block after window expiration may need
> to dedicate most of its space to corrections, temporarily sacrificing
> transaction throughput for safety.

> 

Reviewer 2: :heart_eyes:

<a id="certificate-inclusion" href="#certificate-inclusion"></a>**Certificate
Inclusion**: Block producers creating new RBs include certificates for EBs where
at least $L_\text{vote} + L_\text{diff}$ slots have elapsed since the slot of
the RB that announced the EB (step 12). This timing constraint ensures the
certified EB has had sufficient time to diffuse throughout the network.

**Transaction Selection**: When producing an RB that includes an EB certificate,
the block contains only the certificate. When producing an RB without a
certificate, the producer includes transactions from their mempool only if they
can validate all transactions against the complete ledger state (including all
certified EBs). If missing any certified EB data, the producer creates an empty

> RB.

Reviewer 2: This paragraph is confusing. It switches from passive to active and the conditions are not clear enough to convince it covers all cases.

the potential EB-related work a node must do when demand is high.

<div align="center">
<a name="figure-12" id="figure-12"></a>

![Size of transactions referenced by EBs](images/contents-ebs-size.svg)

<em>Figure 12: Size of transactions referenced by EBs</em>

</div>

Because of the aforementioned backpressure, diffusion occurs in Leios in an
orderly manner even when demand is high. The following set of plots show
histograms of diffusion time (i.e., the time from a transaction's, RB's, EB's,
or vote's creation to its reaching the nodes in the network). Transactions and

> votes typically diffuse rapidly to ~95% of honest stake in fractions of a

Reviewer 2: Where are the 95% coming from here?

> <a name="feasible-parameters"></a>**Feasible protocol parameters**

Reviewer 2: It was very annoying to find this section and it is not really set apart from the above and below (containing) text. We should really consider using normal headings:

```suggestion
### Feasible protocol parameters
```

**Parameter Relationships and Network Assumptions**

> The key innovation in this specification is the relationship between the voting
> threshold ($\tau = 75\%$) and the diffusion period ($L_\text{diff}$). The high

Reviewer 2: ```suggestion
The key relation in the proposed protocol is between the voting
threshold ($\tau = 75\%$) and the propagation delay of EBs ($\Delta_\text{EB}$). The high
```

It's not only L_diff which is at play here and we must balance both L_vote and L_diff with assumed Delta_EB (and Delta_RB!).

voting threshold ensures that any certified EB is already known to at least 25%
of honest nodes by the end of $L_\text{vote}$, even assuming 50% adversarial
stake. This widespread initial knowledge enables the critical network

> assumption: an EB known to >25% of the network will reach all honest parties
> within the additional $L_\text{diff}$ period.

Reviewer 2: ```suggestion
assumption: an EB evidently known to >25% of the network will reach the remaining (online) honest stake within $L_\text{diff}$ and $\Delta_\text{RB}$, so to not impact blockchain safety and liveness.
```

**Example Parameter Calculation**

To illustrate how these relationships translate into concrete parameters,
consider the following example based on simulated network measurements:

**Given Network Characteristics:**

> - $\Delta_\text{hdr} = 1$ slot, $\Delta_\text{RB} = 5$ slots (current Praos
>   observations)

Reviewer 2: ```suggestion
- $\Delta_\text{hdr} = 1$ slot, $\Delta_\text{RB} = 5$ slots (mainnet assumption for Praos security)
```

> - $\Delta_\text{EB}^{\text{opt}} = 7$ slots, $\Delta_\text{EB}^{\text{wc}} = 15$

Author: Happy to change it

  slots (EB diffusion times)
- $\Delta_\text{cpu} = 3$ slots, $\Delta_\text{reapply} = 1$ slot (validation
  times)

**Timing Parameter Calibration:**

- $L_\text{vote} \approx \Delta_\text{EB}^{\text{opt}} = 7$ slots (ensures
  honest EBs diffuse optimally, with margin for validation)
- $L_\text{diff} \geq \Delta_\text{EB}^{\text{wc}} + \Delta_\text{reapply} - \Delta_\text{RB} - L_\text{vote} = 15 + 1 - 5 - 7 = 4$
  slots (minimum)

> - Setting $L_\text{diff} = 8$ slots provides additional safety margin

Reviewer: ```suggestion
- Setting $L_\text{diff} = 7$ slots provides additional safety margin
```

@pagio recommended 7, and we're using that in the simulations. However, 8 is fine for this example.

| RB header diffusion bound                     | $\Delta_\text{hdr}$ |       1 slot       | Must be faster than full RB diffusion for equivocation detection; simulations show headers reach all nodes within 1 slot due to smaller size.                                          |
| Voting period length                          |   $L_\text{vote}$   |      7 slots       | Short stages increase settlement speed, but the stage length must be generously larger than the propagation time for EBs.                                                              |
| Diffusion period length                       |   $L_\text{diff}$   |     5-10 slots     | Calculated as $\Delta_\text{EB}^{\text{wc}} - \Delta_\text{RB} - L_\text{vote}$. With typical values: ~20 - 5 - 7 = 8 slots                                                            |
| Endorser-block referenceable transaction size |  $S_\text{EB-tx}$   |       12 MB        | Simulations indicate that 200 kB/s throughput is feasible at this block size.                                                                                                          |

> | Endorser block max size                       |    $S_\text{EB}$    |       512 kB       | Endorser blocks must not be so large that they cannot diffuse and be validated within the voting period $L_\text{vote}$.                                                               |

Reviewer 2: Double negation. Consider rephrasing for more clarity

> | Maximum Plutus steps per endorser block       |          -          |  2000G step units  | Simulations at high transaction-validation CPU usage.                                                                                                                                  |

Reviewer: We'll need to lower this to 1000 Gstep or 600 Gstep, based on the simulation using `sim-rs-1.2.0`. I'll fix this in a separate PR.
